server-master/src/main/java/com/aliyun/emr/rss/service/deploy/master/clustermeta/ha/HANodeDetails.java

```diff
@@ -160,4 +169,27 @@ public static boolean isLocalAddress(InetAddress addr) {
 
     return local;
   }
+
+  public static boolean isLocalPortOccupied(int port) {
+    ServerSocket ss = null;
+    DatagramSocket ds = null;
+    try {
+      ss = new ServerSocket(port);
```

Comment:
I think it is dangerous to occupy a port to check its availability. And you'll never know what might happen after you release a port.

Reply:
Agree, and may i know is there any other way to check if the port is usable without actually occupying it?

Reply:
Even if we found a way to check port availability here, we can't guarantee that the port will not be occupied before ratis uses it. I think we should check what happens when port is unavailable but ratis server starts, better to throw Exception and exit then.